dspace-api/src/main/java/org/dspace/core/Context.java

```diff
@@ -628,6 +628,17 @@ public boolean inSpecialGroup(UUID groupID) {
         return myGroups;
     }
 
+    /**
+     * This method will remove any special group that was assigned to the context.
+     * This should be used in very specific scenario such as login-as feature where
+     * we don't want to pass special group related to the current request /
+     * authentication method. Normally an authentication method only need to add it
+     * own additional group without touching what is already set in the context
+     */
+    public void emptySpecialGroups() {
+        specialGroups.clear();
+    }
```

Comment:
If I correctly understand the purpose of this new method, I'd  like to suggest we *instead* change the behavior of `setCurrentUser()` to automatically clear the (cached) `specialGroups` List whenever it is called. 

It seems fragile to me to expect code to know when to empty out the specialGroups listing by calling this new method.  Instead, it's more logical to remove the cached specialGroups whenever the current user changes....a new current user should *not* inherit the groups of the last current user.

Reply:
@tdonohue  your understanding is correct and I agree that the method must be used with caution (this is what I try to tell in the javadoc). Your idea is interesting, we cannot do that directly in the setCurrentUser because it is used elsewhere also as part of the authentication process so I think it would be risky to change, what about introducing a specific method `loginAs` (and a `terminateLoginAs`)? this would be easier to understand, reusable and we can encapsulate here the logic to store the current specialGroup, empty and restore later.

BTW: restoring is not really needed as on the REST side each request will be just limited to a single functionality but for the sake of clarity it is just good to "restore" the context before that the request complete

Reply:
If we'd modify that, we have to be careful to verify `setCurrentUser()` is never used where the `specialGroups` need to be retained (e.g. as part of an authentication method), especially since there were no tests for specialGroups (Andrea has created the first test for it in this PR)

Reply:
If we are concerned about changing `setCurrentUser()` at this time, I'm OK with @abollini's idea to have a separate `loginAs` (and perhaps `terminateLoginAs`) methods alongside `setCurrentUser()`.  I just don't like this new `emptySpecialGroups()` method, as it implies that other code *needs to understand the inner workings of the Context class*, and know exactly when specialGroups are cached and when they need emptying.  I'd rather the Context class make those decisions itself, and a `loginAs` and `terminateLoginAs` would align better with that idea.

Reply:
This does sound like a good alternative to me as well, but I would prefer a different name because DSpace offers a login as feature, and that can't work with this method (since it has to be permanent)

Reply:
done, see f0210c5a46bbb5f9123ec1e4cb5187e9c441e590